### DISCUSSION TITLE
Why don't we allow expressions as statements?

### DISCUSSION THREAD

**Discussion by grifx** [ricomariani/CG-SQL-author#20] [Q&A]
```sql
DECLARE PROC printf NO CHECK;

create procedure dump_bool(value bool, file_path text, line_number int, out result bool) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "bool", case value when TRUE then "TRUE" when FALSE then "FALSE" else "NULL" end); end;
create procedure dump_int(value int, file_path text, line_number int, out result int) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "int", case value when null then "NULL" else printf("%d", value) end); end;
create procedure dump_long(value long, file_path text, line_number int, out result long) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "long", case value when null then "NULL" else printf("%lld", value) end); end;
create procedure dump_real(value real, file_path text, line_number int, out result real) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "real", case value when null then "NULL" else printf("%f", value) end); end;
create procedure dump_text(value text, file_path text, line_number int, out result text) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "text", case value when null then "NULL" else printf("%s", value) end); end;
create procedure dump_blob(value blob, file_path text, line_number int, out result blob) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "blob", case value when null then "NULL" else printf("blob") end); end;
create procedure dump_object(value object, file_path text, line_number int, out result object) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "object", case value when null then "NULL" else printf("object") end); end;
#define dump() dump(__FILE__,  __LINE__)

CREATE PROC entrypoint ()
BEGIN
  IF TRUE::dump() THEN
    call printf("true works\n");
  ELSE
    call printf("true doesn't work\n");
  END IF;

  -- "That would be nice"::dump()
END;
```

It could be useful here too:
```sql
CREATE PROC entrypoint ()
BEGIN
  let my_list := create_string_list():::append("john"):::append("patrick");

  IF TRUE THEN
    my_list
      :::append("steve")
      :::append("bob")
    ;
  END IF;
END;
```

**Comment by ricomariani** [comment]
In this context append returns a value, in general every function call does, there are no void expressions.

Do you want this to work?

  begin
     printf("whatever");
  end;

or this?

   list::append("foo");

They aren't the same thing in CQL so there is ambiguity in the grammar.

I might be able to make "call" optional without ambiguity.  I can't do both.

On the other hand, you could imagine some syntax that says to evaluate an expression and discard the result.  But this isn't really a natural thing in CQL.

It's something I've long considered.

C programmers really find "call" offensive... but it makes other syntax possible and symmetric.

Check out the grammar diagram and look at the other places call appears.

The whole thing is very subtle...

R


________________________________
From: Joris Garonian ***@***.***>
Sent: Friday, September 8, 2023 11:51 PM
To: ricomariani/CG-SQL-author ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ricomariani/CG-SQL-author] Why don't we allow expressions as statements? (Discussion #20)


DECLARE PROC printf NO CHECK;

create procedure dump_bool(value bool, file_path text, line_number int, out result bool) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "bool", case value when TRUE then "TRUE" when FALSE then "FALSE" else "NULL" end); end;
create procedure dump_int(value int, file_path text, line_number int, out result int) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "int", case value when null then "NULL" else printf("%d", value) end); end;
create procedure dump_long(value long, file_path text, line_number int, out result long) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "long", case value when null then "NULL" else printf("%lld", value) end); end;
create procedure dump_real(value real, file_path text, line_number int, out result real) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "real", case value when null then "NULL" else printf("%f", value) end); end;
create procedure dump_text(value text, file_path text, line_number int, out result text) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "text", case value when null then "NULL" else printf("%s", value) end); end;
create procedure dump_blob(value blob, file_path text, line_number int, out result blob) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "blob", case value when null then "NULL" else printf("blob") end); end;
create procedure dump_object(value object, file_path text, line_number int, out result object) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "object", case value when null then "NULL" else printf("object") end); end;
#define dump() dump(__FILE__,  __LINE__)

CREATE PROC entrypoint ()
BEGIN
  IF TRUE::dump() THEN
    call printf("true works\n");
  ELSE
    call printf("true doesn't work\n");
  END IF;

  -- "That would be nice"::dump()
END;

It could be useful here too:

CREATE PROC entrypoint ()
BEGIN
  let my_list := create_string_list():::append("john"):::append("patrick");

  IF TRUE THEN
    my_list
      :::append("steve")
      :::append("bob")
    ;
  END IF;
END;

**Comment by ricomariani** [comment]
I tried implied calls and that works but I'm not sure it's wise.

In CQL there's a difference between a procedure call an a function call.  A procedure can use the database, a function can't.  A procedure doesn't have a result except maybe out arguments.  A procedure can fail a function can't.  So like

call foo();   -- this could throw

if fundamentally different than

set v := foo();  -- this can't...

There is one except to this, for convenience you can call a procedure like it's a function using its final out value as the return.  This is a bit of convenience mostly for procs that can't fail anyway.

create proc fib(n integer not null, out result integer not null)
begin
  set  result := case when <= 2 then 1 else fib(n-1)+fib(n-2) end;
end;

That works because of "proc as func".

But it rapidly gets weird for "real" procedures.

call insert_stuff_into_db(); -- this can fail

is very different than

var result := insert_item_into_list(list, 7);  -- does not throw...

Maybe "proc as func" was a mistake but it is highly useful and it's been in the code since the earliest times.  Sigh.

I think it's times like this that I just think.  "Stop trying to make it into C, it isn't C..." and then I sleep better 😄

R



________________________________
From: Joris Garonian ***@***.***>
Sent: Friday, September 8, 2023 11:51 PM
To: ricomariani/CG-SQL-author ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ricomariani/CG-SQL-author] Why don't we allow expressions as statements? (Discussion #20)


DECLARE PROC printf NO CHECK;

create procedure dump_bool(value bool, file_path text, line_number int, out result bool) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "bool", case value when TRUE then "TRUE" when FALSE then "FALSE" else "NULL" end); end;
create procedure dump_int(value int, file_path text, line_number int, out result int) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "int", case value when null then "NULL" else printf("%d", value) end); end;
create procedure dump_long(value long, file_path text, line_number int, out result long) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "long", case value when null then "NULL" else printf("%lld", value) end); end;
create procedure dump_real(value real, file_path text, line_number int, out result real) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "real", case value when null then "NULL" else printf("%f", value) end); end;
create procedure dump_text(value text, file_path text, line_number int, out result text) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "text", case value when null then "NULL" else printf("%s", value) end); end;
create procedure dump_blob(value blob, file_path text, line_number int, out result blob) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "blob", case value when null then "NULL" else printf("blob") end); end;
create procedure dump_object(value object, file_path text, line_number int, out result object) begin set result := value; call printf("[DUMP] (%s:%d): (%s) %s\n", file_path, line_number, "object", case value when null then "NULL" else printf("object") end); end;
#define dump() dump(__FILE__,  __LINE__)

CREATE PROC entrypoint ()
BEGIN
  IF TRUE::dump() THEN
    call printf("true works\n");
  ELSE
    call printf("true doesn't work\n");
  END IF;

  -- "That would be nice"::dump()
END;

It could be useful here too:

CREATE PROC entrypoint ()
BEGIN
  let my_list := create_string_list():::append("john"):::append("patrick");

  IF TRUE THEN
    my_list
      :::append("steve")
      :::append("bob")
    ;
  END IF;
END;

**Comment by ricomariani** [comment]
What we could do that doesn't break the grammar:
  * we could a statement that is a call with the `call` keyword absent (I tried this it creates no ambiguity)
  * we can make it so that you can "call" a function and hence implicitly discard the result

This would be sort of the reverse of proc_as_func (func_as_proc).   If we did then then you would get procedure calls as statements with no keyword and that's really the most important thing.  This avoids the grammar problems that allowing any expression causes but it does make it less obvious what can throw and what can't.  Also the cursor forms still would need "call" but that's probably ok.

As a consequence:  `call  your_function(args);`  would also work and discard the result.

This turns out to be not totally trivial, but it is doable.

Note we would also have to support :, ::, and ::: in the context of a CALL to get all the things you wrote above.  This is also doable.

**Reply by grifx** [reply]
Just to make sure we are on the same page. Can you confirm this would work?

```
IF TRUE::dump() THEN
  FALSE::dump();
  "This is a test"::dump();
END IF
```

(The idea of the dump util is to stdout any value wherever it is.)

---

If you have spare time can you expand on why the grammar would be broken when allowing any expression as statement?

```
TRUE;
TRUE::dump();
FALSE;
"Bla";
```

I understand that it's dirty and the intent ambiguous but I don't see the technical constraint. Not pushing this idea forward, just out of curiosity.

Maybe in the future some errors could be warnings and warnings could be silenced?
For example, sometimes it might be hard to implement macros when it's too restrictive.

**Comment by ricomariani** [comment]
FALSE::dump() is an expression so that doesn't work.

I had fmt returning a string rather than printing to make this clear.

Now with regard to this expression statement business.  The last time I tried this it was a catastrophe so my plan was to just add the line and show you all the stuff that blows up.  But then ... nothing blowed up.  I think one of my grammar refactors previously made it unambiguous again but I'm not sure which one.

So I have a stash that parses

expr_stmt: expr

and then adds

expr_stmt to the any_statement list

I could do codegen for this without much trouble.

This would almost do everything we need.  I'd have to make it so that you can call a procedure using function syntax for it to work out fully but that seems doable.

So maybe I can do this after all?


________________________________
